SIPS/sip-29.md

```diff
@@ -0,0 +1,232 @@
+type OnAssetLookupReturn = {
+  assets: Record<Caip19AssetType, AssetMetadata>;
+};
+```
+
+#### Get Asset Conversion Rate
+
+```typescript
+import { OnAssetConversionHandler } from "@metamask/snaps-sdk";
+
+export const onAssetConversion: OnAssetConversionHandler = async ({
+  conversions
+}) => {
+  const conversionRates = /* Get conversion rate */;
+  return { conversionRates };
+};
+```
+
+The type for an `onAssetConversion` handler function’s arguments is:
+
+```typescript
+type Conversion = {
+  from: Caip19AssetType;
+  to: Caip19AssetType;
+};
+
+type OnAssetConversionArgs = {
+  conversions: Conversion[];
+};
+```
+
+The type for an `onAssetConversion` handler function’s return value is:
+
+```typescript
+type AssetConversionRate = {
+  // The rate of conversion from the source asset to the target asset. It
+  // means that 1 unit of the `from` asset should be converted to this amount
+  // of the `to` asset.
+  rate: string;
+
+  // The UNIX timestamp of when the conversion rate was last updated.
+  conversionTime: number;
+
+  // The UNIX timestamp of when the conversion rate will expire.
+  expirationTime: number;
+};
+
+type FromAsset = Conversion["from"];
+
+type ToAsset = Conversion["to"];
+
+type OnAssetConversionReturn = {
+  conversionRates: Record<From, Record<To, AssetConversionRate>>;
+};
+```
+
+### Fiat currency representation
+
+We SHOULD use CAIP-19 to represent fiat currencies as well. This approach
+provides a consistent way to represent all assets, making the API more
+predictable. The proposed format is:
+
+```
+asset_type:        chain_id + "/" + asset_namespace + ":" + asset_reference
+chain_id:          namespace + ":" + reference
+namespace:         "fiat"
+reference:         country_code
+asset_namespace:   "currency"
+asset_reference:   currency_code
+```
+
+The country code is a two-letter lowercase code, as defined by ISO 3166-1
+alpha-2, representing the emitter country, with the exception of the European
+Union, which is represented by "eu".
+
+The currency code is a three-letter uppercase code as defined by ISO 4217.
+
+Examples:
+
+```
+# Euro
+fiat:eu/currency:eur
+
+# United States Dollar
+fiat:us/currency:usd
+
+# Brazilian Real
+fiat:br/currency:brl
+
+# Japanese Yen
+fiat:jp/currency:jpy
+```
+
+## Appendix I: Fungible Asset Metadata
+
+The following asset metadata fields for a fungible asset are defined.
+As of the time of creation of this SIP, they are the only possible assets requested by clients.
+
+```typescript
+// Represents an asset unit.
+type FungibleAssetUnit = {
+  // Human-friendly name of the asset unit.
+  name: string;
+
+  // Ticker of the asset unit.
+  ticker: string;
+
+  // Number of decimals of the asset unit.
+  decimals: number;
+};
+
+// Fungible asset metadata.
+type FungibleAssetMetadata = {
+  // Human-friendly name of the asset.
+  name: string;
+
+  // Ticker of the asset.
+  ticker: string;
+
+  // Whether the asset is native to the chain.
+  native: boolean;
```

Comment:
> How can we identify other types of tokens that need to be handled differently?

Reply:
Some examples might be differentiating between brc-20 and rune tokens on bitcoin. Or eth, erc20s, and staking contracts on ethereum.  I anticipate more types of tokens than just native vs not. Not sure if anything built into caip19 already helps distinguish this.